### ISSUE TITLE
Understanding the logic behind giving custom pairs for pairwise logit training

### ISSUE THREAD

**Issue by anuragreddygv323** [catboost/catboost#1063]
Problem: In the context of using custom pairs for pairwise ranking, How does catboost model knows the relevance for pairs when I create a pool object using 

pool1= Pool(data=X_train, label=y_train, group_id=group_ids, pairs=pairs)

In pairs = pairs parameter, if a pair is 0,1 : it means that 0th index and 1st index from X_train will be taken as features  and y_train 0th index and 1st index is taken as the relevance labels respectively and pairwise logit is computed?  is it how its computed in the backend ?

Please verify my understanding...

**Comment by anuragreddygv323** [comment]
@annaveronika

**Comment by annaveronika** [comment]
If pairs are provided, then label values are ignored when calculating the loss function. Only the pairs that are provided are used for loss calculation. Loss is calculated as described in calculation principles of PairLogit and PairLogitpairwise here: https://catboost.ai/docs/concepts/loss-functions-ranking.html

**Comment by annaveronika** [comment]
Let us know, if you have more questions! Closing the issue.

**Comment by anuragreddygv323** [comment]
pool1= Pool(data=X_train, label=y_train, group_id=group_ids, pairs=pairs)

In this, how will catboost know what features to pick for each pair? is it based on how the X_train is arranged( index based ?)

**Comment by anuragreddygv323** [comment]
If pairs are provided, then label values are ignored when calculating the loss function. 

Do you mean that the label values are ignored for computing the pairs ? 

We need the labels for the pairs to compute the loss right ?
pool1= Pool(data=X_train, label=y_train, group_id=group_ids, pairs=pairs)

**Comment by annaveronika** [comment]
If pairs are provided, pairs are not computed, the ones that are provided by the user are used for optimization. We don't need labels to compute the loss if the user provides pairs.

**Comment by anuragreddygv323** [comment]
pool1= Pool(data=X_train, label=y_train, group_id=group_ids, pairs=pairs)

The pool object requires labels right ?


So we minimize the negative loglikelihood:

−∑𝑖,𝑗∈𝑃𝑎𝑖𝑟𝑠log(11+exp−(𝑓(𝑑𝑖)−𝑓(𝑑𝑗)))
 
We need labels for getting this summation which we are trying to minimize. I did not follow when u said "We don't need labels to compute the loss if the user provides pairs" 

Could you elaborate... ?

**Comment by annaveronika** [comment]
The pool object doesn't require labels, labels are not used in summation, the summation uses all pairs provided in pairs parameter. To calculate the loss we need to know the list of pairs, for each pair we calculate loss which uses formula values on each object (f(di)) and the fact that di is the winner, this fact is stated in the pair, see documentation for pairs parameter:

The first element of the pair is the zero-based index of the winner object from the input dataset for pairwise comparison.
The second element of the pair is the zero-based index of the loser object from the input dataset for pairwise comparison.
